clash-prelude/src/Clash/XException/MaybeX.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE PatternSynonyms #-}
+
+-- | Helpers to make 'Clash.XException.XException' explicit in the type system.
+-- Using these helpers can help programmers account for 'Clash.XException.XException's
+-- properly in blackbox models or tests.
+module Clash.XException.MaybeX
+  ( MaybeX(IsX, IsDefined)
+
+    -- * Construction
+  , toMaybeX
+
+    -- * Deconstruction
+  , unsafeToA
+
+    -- * Operations
+  , andX
+  , orX
+  , maybeX
+  ) where
+
+import Prelude
+
+import Control.Applicative
+import Clash.XException (errorX, isX)
+
+-- | Structure helping programmers to deal with 'Clash.XException.XException'
+-- values. For safety reasons it can't be constructed directly, but should be
+-- constructed using either 'pure' or 'toMaybeX'. After construction, it can be
+-- deconstructed using either 'IsX' or 'IsDefined'.
+data MaybeX a
+  = IsX_ String
+  -- ^ Upon construction, @a@ evaluated to 'Clash.XException.XException'
+  | IsDefined_ !a
+  -- ^ Upon construction, @a@ evaluated to a non-bottom WHNF
+  deriving (Show)
```

Comment:
Maybe it's a good idea to write a custom `Show` instance that uses the exported patterns `IsX` and `IsDefined`?

Reply:
Yep